reverseproxy: Fix dynamic upstreams ip version

armadi1809 opened the pull request:
Closes #6442

francislavoie commented on modules/caddyhttp/reverseproxy/upstreams.go:
```diff
@@ -313,8 +313,8 @@ func (au *AUpstreams) Provision(ctx caddy.Context) error {
 
-	resolveIpv4 := au.Versions == nil || au.Versions.IPv4 == nil || *au.Versions.IPv4
-	resolveIpv6 := au.Versions == nil || au.Versions.IPv6 == nil || *au.Versions.IPv6
+	resolveIpv4 := au.Versions == nil || (au.Versions.IPv4 == nil && au.Versions.IPv6 == nil) || (au.Versions.IPv6 != nil && *au.Versions.IPv4)
```
```suggestion
	resolveIpv4 := au.Versions == nil || (au.Versions.IPv4 == nil && au.Versions.IPv6 == nil) || (au.Versions.IPv4 != nil && *au.Versions.IPv4)
```

francislavoie commented:
I think a test would be useful for this if you have time to look into that 🤔 we could refactor this logic to a separate function to return a tuple of booleans so it can be tested as a unit.

armadi1809 commented:
@francislavoie I just refactored the logic determining the ip version into its own function and added a test covering all possible combinations of `au.Versions`.

francislavoie commented on modules/caddyhttp/reverseproxy/upstreams.go:
```diff
@@ -397,6 +386,19 @@ func (au AUpstreams) GetUpstreams(r *http.Request) ([]*Upstream, error) {
 
 func (au AUpstreams) String() string { return net.JoinHostPort(au.Name, au.Port) }
 
+func resolveIpVersion(versions *IPVersions) string {
```
I think we should move this function to just below the IPVersions type (higher in this file) since it's tightly related.

armadi1809 replied:
Done.

francislavoie reviewed (approved):
Thanks! Looks great 😊

calvinbrewer commented:
Still running into this issue, but it looks like a release hasn't been cut since June 2nd. What is the release cadence for deploying these new changes? Cheers!
